### ISSUE TITLE
Error: There isn't a common type to coerce Binary and Utf8 in LIKE expression

### ISSUE THREAD

**Issue by JayjeetAtGithub** [apache/arrow-datafusion#7342]
### Describe the bug

On running the 4 queries below on the Clickbench multi file dataset, 

```
SELECT COUNT(*) FROM hits WHERE "URL" LIKE '%google%';

SELECT "SearchPhrase", MIN("URL"), COUNT(*) AS c FROM hits WHERE "URL" LIKE '%google%' AND "SearchPhrase" <> '' GROUP BY "SearchPhrase" ORDER BY c DESC LIMIT 10;

SELECT "SearchPhrase", MIN("URL"), MIN("Title"), COUNT(*) AS c, COUNT(DISTINCT "UserID") FROM hits WHERE "Title" LIKE '%Google%' AND "URL" NOT LIKE '%.google.%' AND "SearchPhrase" <> '' GROUP BY "SearchPhrase" ORDER BY c DESC LIMIT 10;

SELECT * FROM hits WHERE "URL" LIKE '%google%' ORDER BY to_timestamp_seconds("EventTime") LIMIT 10;

```

we get this error,

```bash
type_coercion caused by Error during planning: There isn't a common type to coerce Binary and Utf8 in LIKE expression
```

### To Reproduce

Download the data using,

```bash
 ./benchmarks/bench.sh data clickbench_partitioned
```

A `hits_multi` directory with the parquet files will be created.

Execute the above queries,

```bash
datafusion-cli -c "CREATE EXTERNAL TABLE hits STORED AS PARQUET LOCATION 'hits_multi';" "{query}"
```

### Expected behavior

The queries should run successfully without erroring.

### Additional context

Datafusion 29.0.0

**Comment by alamb** [comment]
This looks similar to https://github.com/apache/arrow-datafusion/issues/7039 which @jonahgao  fixed by adding a coercion from binary --> UTF8 for comparison. I think we could do something similar here.

**Comment by alamb** [comment]
Marking as a good first issue as there is a reproducer and I think the fix should be relatively straightforward

**Comment by JayjeetAtGithub** [comment]
Working on a fix

**Comment by JayjeetAtGithub** [comment]
I looked into this issue a little bit. Looks like there needs to be changes in `arrow-string` which is basically a part of `arrow-rs`. Specifically, I found out that there is [no support for `like` operation for binary operands](https://github.com/apache/arrow-datafusion/blob/ae4b52a7397bd7f613ea3782a84cc27245d0d57d/datafusion/physical-expr/src/expressions/like.rs#L23-L28) in `arrow-datafusion`. It only exists for `utf8` operands. Support for the `like` operation in `utf8` stems from [this](https://github.com/apache/arrow-rs/blob/90449ffb2ea6ceef43ce8fc97084b3373975f357/arrow-string/src/like.rs#L337) code in `arrow-string`. We basically need to add a binary version of this function in `arrow-rs`. Does that sound correct ?
